hkmc2/shared/src/main/scala/hkmc2/semantics/flow/FlowAnalysis.scala

```diff
@@ -70,18 +82,17 @@ class FlowAnalysis(using tl: TraceLogger)(using Raise, State, Ctx):
         case stmt: DefineVar =>
           val rhs = typeProd(stmt.rhs)
           stmt.sym match
-          case sym: FlowSymbol =>
-            constrain(rhs, C.Flow(sym))
+          case sym: FlowSymbol => constrain(rhs, C.Flow(sym))
+          case _ => ()
         case t: TermDefinition =>
-          val sign_ty = t.sign.map(typeProd) // TODO use sign_ty
+          val sign_ty = t.sign.map(typeProd(_, insideSelAppChain)) // TODO use sign_ty
           val ps = t.params.map(typeParamList)
           t.body.foreach: bod =>
             val bod_ty = typeProd(bod)
             val fun_ty = ps.foldRight(bod_ty): (pl, acc) =>
               P.Fun(C.Tup(pl, N), acc, Nil)
             constrain(fun_ty, C.Flow(t.sym.flow))
-        case t: Term =>
-          typeProd(t)
+        case t: Term => typeProd(t)
```

Comment:
You have a lot of useless changes like that. In bit pattern matches, I find this form less readable. Please revert all of them.